docs/lab/02/02.md

```diff
@@ -0,0 +1,223 @@
+
+**Input**
+```rust
+let gpio_name = Input::new(p.PIN_n, Pull::Up);
+```
+
+For the input type pins, as well as the output ones, we must set the pin number in `PIN_n`, `n` being the number in the pin diagram. `Pull::Up` or `Pull::Down` activates the internal pull up or pull down resistors of the GPIO pin.
+
+:::warning 
+
+For a correct use of the buttons, use pull-up, pull-down resistors depending on the mode of operation of the button. Check [01 - Hardware Introduction](https://embedded-rust-101.wyliodrin.com/docs/lab/01#buttons)
+
+:::
+
+## Control GPIO Pins
+
+As we saw in the previous chapters, the GPIO pins can be used to communicate with the environment through external elements. Next we will see how we can do this using Embassy-rs.
+
+### Writing Rust code to control LEDs using GPIO pins.
+
+The LEDs can be controlled by gpio because the pins can output electric current depending on their configuration (read the datasheet). Therefore, after you configure the pin to correspond with your led, it is necessary to control the voltage for the respective led, so we have to look at the panel of the GPIO pin and choose the appropriate resistance.
+
+**Steps to turn on/off a led:**
+
+1. First of all, we will define our LED and assign it the Level::Low property to be initially off.
+```rust
+let mut led = Output::new(p.PIN_17, Level::Low);
+```
+
+2. To control the flow of current passing through the pin, we have 2 functions to set the pin to low or high depending on the need.
+```rust
+led.set_high();
+led.set_low();
+```
+
+:::warning 
+
+The GPIO pins output a standard output current depending on the microcontroller. To be sure that the LED will work normally and there is no risk of destruction, it is recommended to calculate and add a resistor to control the current level.
+
+:::
+
+<div align="center">
+![LED GPIO](images/led-gpio17.png)
+</div>
+
+### Writing Rust code to read button state using GPIO pins.
+
+To read the input from the button, we must define the gpio we will use and also choose whether we will put a pull-up or pull-down resistor.
+
+**Steps to turn read state of a button:**
+
+1. We need to define the pin used and set the type of resistance used.
+```rust
+let button = Input::new(p.PIN_2, Pull::Up);
+```
+
+2. To read the input value we can use the is_high() function which will return true if the pin detects input current or false if the pin does not detect input current.
+```rust
+if button.is_high() {
+    // Do something ...
+} else {
+    // Do something ...
+}
+```
+
+<div align="center">
+![Button GPIO](images/gpio-button.png)
+</div>
+
+## Advanced techniques for using GPIO pins
+
+In this topic we will discuss several techniques used to solve some hardware problems that can appear using only software.
+
+### Using async for efficient GPIO handling.
```

Comment:
Please do this with a loop, reading continuously. We did not discuss async/await yet.